File: 0966-binary-subarrays-with-sum/0966-binary-subarrays-with-sum.cpp
class Solution {
public:
    int numSubarraysWithSum(vector<int>& nums, int goal) {
        unordered_map<int,int> mp; //prefixsum,count
        mp[0]=1;
        int prefixSum=0;
        int ans=0;
        for(auto num:nums) {
            prefixSum+=num;
            int diff = prefixSum-goal;
            if(mp.find(diff)!=mp.end()) ans+=mp[diff];
            mp[prefixSum]++;
        }
         return ans;

    }
};

File: 0966-binary-subarrays-with-sum/0966-binary-subarrays-with-sum.py
class Solution:
    def numSubarraysWithSum(self, nums: List[int], goal: int) -> int:
        def atMostKSum(k):
            currentSum = 0
            answer = 0
            l = 0
            for r,num in enumerate(nums):
                currentSum+=num
                while l<=r and currentSum>k:
                    currentSum-=nums[l]
                    l+=1
                if currentSum<=k:
                    answer +=(r-l+1)
            return answer
        
        return atMostKSum(goal)-atMostKSum(goal-1)
        